### ISSUE TITLE
Wayland

### ISSUE THREAD

**Issue by TH3-S4LM0N** [leftwm/leftwm#1035]
### Describe the feature

There was some talk about Wayland support in the discord, and I've personally been looking into a fair bit. It just makes more sense to do something formal on the actual repo. I also just wanted to make a general info dump somewhere and a sort of todo list. I'm also making this as it's a lot of work and I for sure can't do the whole thing myself.
In my opinion what needs to happen for Wayland support is this (completely up for debate, what the issue is for):
- Everything needs to be reworked into 1 binary, mainly for better state things (eg if it's running wl or x) so that all that doesn't have to be passed around all the different binaries. (This may also cause a rework of `leftwm command` as it can just call a function instead of having to write the command to a state file). 
- A more modular design. While the `DisplayServer` trait is a great idea, how LeftWM is launched doesn't handle modularity well. My general idea is:
    - Cargo features for `xorg` and `wayland`.
    - CLI args to pick between them at launch (`start` subcommand).
I've already done some of the work for this but other eyes would be nice.

After all that there's the obvious task of actually implementing Wayland support. My understanding is that there are 2 options for how to interface with Wayland:
- [Smithay](https://github.com/Smithay/smithay). A pure Rust connection to Wayland designed for making compositors.
- [`wayland-client`](https://github.com/Smithay/wayland-rs). The backend for Smithay, providing more general connections to the Wayland server. Also pure Rust.

In my experience after trying both I would say Smithay is too based around its own event loop and being the integral part of compositors built from the ground up with it. `wayland-client` seems to me to be more what LeftWM is looking for as an already established project.

Of course this will break pretty much everything related to LeftWM besides LeftWM itself (themes that use `leftwm-*`, all packages).
A `wayland` branch or something would be nice once there's actual code to be pr'd.

A policy or something about Nvidia is something to be thought through, as I'm sure yall've heard the horror stories, whether that be no Nvidia support or special features.

As of my answers to the questions below, I would say LeftWM will be completely irrelevant with ~6 months without Wayland as X is dead or on life support depending on how you look at it.

### How important is this feature to you?

LeftWM is unusable without it

### How important do you think is this feature for the average LeftWM user?

No one uses LeftWM because this is missing

**Comment by Eskaan** [comment]
> This may also cause a rework of leftwm command as it can just call a function instead of having to write the command to a state file.

I don't think so. You would still have one daemon process that is your window manager, and then start a new process by calling the subcommand. The state socket is then used to communicate between the two processes, which brings me to one of my ideas:
Rework the code so it uses dbus instead of a manual state file. To be honest, I haven't looked into this idea as much atm, but I think what we do with our state file is exactly that dbus does.

Also, I am against squishing all of leftwm into one single binary. I propose two binaries like `leftwm` (the WM) and `left-tools` (all other utilities).

Now that I think about it, we could also include `leftwm-theme` via a git submodule and a cargo feature.

I think you can just go and create a pr on a not yet existing branch; it will be created upon merge.

**Comment by 0323pin** [comment]
> I would say LeftWM will be completely irrelevant with ~6 months without Wayland as X is dead or on life support depending on how you look at it.

Latest release was on 2023-02-07 and the one prior to that was 2022-12-19. The only issue I have with Wayland is `wlroots` being written with Linux only in mind.
Wayland is far from being close to usable on non-Linux systems. Right now LeftWM runs flawless on non-Linux systems (*BSD and even MacOS through `xquartz`).

So, by all means bring Wayland support to the table but, please do not kill X straight away.

**Comment by TH3-S4LM0N** [comment]
> The only issue I have with Wayland is `wlroots` being written with Linux only in mind. Wayland is far from being close to usable on non-Linux systems. Right now LeftWM runs flawless on non-Linux systems (*BSD and even MacOS through `xquartz`).

We aren't using `wlroots` as there are no maintained bindings anymore. That's just a general Wayland issue.

> So, by all means bring Wayland support to the table but, please do not kill X straight away.

I never said anything about killing X off, there's no reason to delete stuff that works. X isn't exactly getting breaking updates anymore so there shouldn't have to be almost any work done to maintain the X part of LeftWM, but Wayland should be the new focus. For Linux though (which I would say it's safe to assume is most users) what I said stands.

**Comment by TH3-S4LM0N** [comment]
> which brings me to one of my ideas: Rework the code so it uses dbus instead of a manual state file. To be honest, I haven't looked into this idea as much atm, but I think what we do with our state file is exactly that dbus does.

Actually this is a pretty good idea, as along with making all the `leftwm-*` more centralized with how they do things (or lack there of as leftwm itself is doing the things) it could also lead to some possible bspwm ipc type stuff (i get it's not the same thing but there are interesting possibilities).

> Also, I am against squishing all of leftwm into one single binary. I propose two binaries like `leftwm` (the WM) and `left-tools` (all other utilities).

Wouldn't that be 3 though: `leftwm` (the wm), `lefthk` & `left-tools` (everything that's not leftwm or lefthk).

**Comment by hertg** [comment]
> As of my answers to the questions below, I would say LeftWM will be completely irrelevant with ~6 months without Wayland as X is dead or on life support depending on how you look at it (on Linux).

That seems like an overstatement, X will still be prevalent for years to come, for better or worse.

---

While I personally do not have any experience with Wayland whatsoever, I'm very much looking forward to getting Wayland support into LeftWM and am glad if this gets some push. There are however quite a lot of preparations necessary, namely to abstract away most of the logic in a display-server agnostic manner. LeftWM to this day still has quite some annoyances and re-occurring bugs that it shouldn't have, which is also why there is still no `1.0` release. My assesment is that there are some things we need to get in order first, by **iteratively** refactor parts of the code and properly isolate unrelated parts from each other. 

There will be no Wayland Support in 6 months. It's just too big of a task, and a big-bang approach to refactor everything at once is unlikely to succeed. But we will be much closer to Wayland-Support in 6 months if we gradually improve the codebase and simultaneously build up a proper abstraction layer between WM logic and the display-server. We can do both at once, getting closer to Wayland Support, while still continuously improving LeftWM on X.

Also, duplicate of #71

**Comment by TH3-S4LM0N** [comment]
> That seems like an overstatement, X will still be prevalent for years to come, for better or worse.

Although that is admittedly my somewhat exaggerated guesstimate, I did not say X was going to magically break within 6 months. What I'm saying is that X isn't getting any newer, and by extent LeftWM isn't either. While you'll totally be able to keep adding more features to LeftWM, it's not because an X update opened up new possibilities. Until LeftWM moves to Wayland, LeftWM is tied to the past, and only getting better because it doesn't yet have 'full featureset'.

I am aware of #71 but that was more of a "will yall eventually do wayland" rather than a "ok lets do wayland now".

**Comment by mike-lloyd03** [comment]
> No one uses LeftWM because this is missing

I'm in the opposite boat. I'm still on X _because_ of LeftWM. I tried Sway and it... just wasn't LeftWM. But with more and more distros shipping with Wayland DEs, I doubt there will be a lot of people wanting to try LeftWM if it means switching to X11.

I'd love to contribute to this effort. I'm fairly comfortable with Rust but display servers and compositors are way beyond me. But I'd love to participate in testing if/when we get there.

**Comment by 0323pin** [comment]
> We aren't using wlroots as there are no maintained bindings anymore.

It doesn't really make any difference, it will still be bound to use `udev`, `libinput` and most probably `logind` :(

> For Linux though (which I would say it's safe to assume is most users) what I said stands.

This is disappointing position and not a valid reason.

> I never said anything about killing X off, there's no reason to delete stuff that works.

Good :)

**Comment by nurmohammed840** [comment]
If considering rewriting this project with Wayland support... count me in

**Comment by D-Brox** [comment]
> Wayland is far from being close to usable on non-Linux systems.

That's actually a common misconception. Wayland is just a protocol, so as long as a compositor implements it natively, Wayland could theoretically work on any OS.

For example, there are many stable ports of wayland compositors on both FreeBSD and DragonFly BSD that have been working well for years now, and it's even on FreeBSD [documentation](https://docs.freebsd.org/en/books/handbook/wayland/) now. I believe I read somewhere NetBSD and OpenBSD also had basic support, but I'm not sure.

So LeftWM with Wayland could work on BSD, at least on FreeBSD and DragonFly BSD at the moment, and more *BSDs in the future. Heck, it could work on windows with [WLSg](https://github.com/microsoft/wslg)'s wayland support.

**Comment by 0323pin** [comment]
> That's actually a common misconception.

Yes and no.

Yes, here's my laptop running NetBSD `velox` (`swc`, no `wlroots`) on wayland 3 years ago.

![20200225-104931](https://user-images.githubusercontent.com/90570748/224895949-cbd6ef22-17e9-4d7e-9909-3e99580a3ed2.jpg)

No, not really, https://blog.netbsd.org/tnf/entry/wayland_on_netbsd_trials_and

> stable ports of wayland compositors on both FreeBSD and DragonFly BSD

Not necessarily, Dragonfly uses FreeBSD ports as base, just because it's there doesn't mean it works. It could, I haven't tested but, it's not a guarantee.

> So LeftWM with Wayland could work on BSD

For sure, but a lot of work is required. Right now, if I use wayland, I'm stuck on US keyboard layout, thanks `libinput`, which is far from ideal as my keyboard is not US.

So basically, somethings work already but, it's miles away from being usable.

**Comment by VuiMuich** [comment]
Correct me if I am wrong, but as far as I understood wayland last time I dug into this topic, lots of the functionality that in X11 the WM handles is treated as display server territory in Wayland. So I guess we need to bisect the `leftwm-core` crate and move such things to the `displayserver` crate.
And then I guess it would make more sense to start the `leftway` compositor as its own crate that relies on `leftwm-core`, `lefthk`, `leftwm-layout` and possible other crates we might start as we proceed to further split the core into libs.

As others stated, I don't see X die anytime soon, so I also don't see the X11 side of leftwm getting abandoned in favor of wayland either. Of course providing support to users on both display server sides will become tricky at some point, but I guess this is just a chance to grow the number of people being competent enough to help out new users on both ends.

**Comment by stevensonmt** [comment]
Is there a bird's eye view architecture (current and anticipated) of leftwm documented anywhere? If not it might be worth putting something together so that implementation of alternative backends could be easier to reason about.

**Comment by Eskaan** [comment]
I don't think there is anything like that atm, but this is the basic crate layout:
```
              +-< <any display server> <-+
              +--< xlib-display-server <-+
              |                          |
leftwm-core <-+------------<<------------+-< leftwm
                                         |
                                lefthk <-+
```

The display servers are already separated by a crate from leftwm-core, and where currently one display server is, there can of course be more added and switched to by a feature (or similar).

**Comment by stevensonmt** [comment]
> Correct me if I am wrong, but as far as I understood wayland last time I dug into this topic, lots of the functionality that in X11 the WM handles is treated as display server territory in Wayland. So I guess we need to bisect the leftwm-core crate and move such things to the displayserver crate.

Is this necessarily the best approach if X11 is still going to be not only supported but the default? If X11 doesn't consider it displayserver stuff maybe there needs to be a trait for those functionalities that are assumed into Wayland as displayserver and then have the leftway (or whatever backend, maybe even leftx11) implement both displayserver and "other window managery stuff" traits? Leaving `core` as mostly the integration of the two?

**Comment by TH3-S4LM0N** [comment]
> This is me. I've wanted to use Sway and Hyprland but they just don't work the way I want.

And that's why LeftWM needs Wayland (or the other way around). For your average user coming from say GNOME that was shipped by default on their distro, Sway is i3 (hardcore wm stuff) and not especially approachable (i still find i3 intimidating). Sway is really the only "stable" Wayland wm currently but it isn't approachable so people end up w/ something like awesome on X or just go back to GNOME.

**Comment by sfrique** [comment]
> > This is me. I've wanted to use Sway and Hyprland but they just don't work the way I want.
> 
> And that's why LeftWM needs Wayland (or the other way around). For your average user coming from say GNOME that was shipped by default on their distro, Sway is i3 (hardcore wm stuff) and not especially approachable (i still find i3 intimidating). Sway is really the only "stable" Wayland wm currently but it isn't approachable so people end up w/ something like awesome on X or just go back to GNOME.

hyprland is quite stable for a while.

I would like to see leftwm support wayland as well!

**Comment by FelixSchausberger** [comment]
Is there any update on this matter?

**Comment by Eskaan** [comment]
You can look at #1073 for the current state of the Wayland implementation. Any help is greatly appreciated.
